core/src/test/scala/unit/kafka/server/BrokerLifecycleManagerTest.scala

```diff
@@ -219,15 +225,16 @@ class BrokerLifecycleManagerTest {
       Collections.emptyMap(), OptionalLong.empty())
     poll(ctx, manager, registration)
 
+    def nextHeartbeatDirs(): Set[String] =
+      poll(ctx, manager, prepareResponse[BrokerHeartbeatRequest](ctx, new BrokerHeartbeatResponse(new BrokerHeartbeatResponseData())))
+        .data().offlineLogDirs().asScala.map(_.toString).toSet
+    assertEquals(Set.empty, nextHeartbeatDirs())
     manager.propagateDirectoryFailure(Uuid.fromString("h3sC4Yk-Q9-fd0ntJTocCA"))
+    assertEquals(Set("h3sC4Yk-Q9-fd0ntJTocCA"), nextHeartbeatDirs())
     manager.propagateDirectoryFailure(Uuid.fromString("ej8Q9_d2Ri6FXNiTxKFiow"))
+    assertEquals(Set("h3sC4Yk-Q9-fd0ntJTocCA", "ej8Q9_d2Ri6FXNiTxKFiow"), nextHeartbeatDirs())
```

Comment:
Thanks for carefully thinking this through. 

I believe there's a mistake in that sequence of events though:

> nextHeartbeatDirs() causes the time to advance quickly and passes 100

This isn't how `BrokerLifecycleManagerTest#poll()` works. It will only advance time if there is no event in the queue that can immediately run, so o1 will always run before c2.

> o1 is processed at manager.eventQueue and a CommunicationEvent c3 is appended to manager.eventQueue

Because o1 runs before c2, the EventQueue will override c2 with c3 as they're both deferred events with the same tag.

After running this test tens of thousands of iterations, I was unable to reproduce a flaky run.

Reply:
Thanks for the explanation. You are right. So, this is not an issue.